### PR TITLE
🌱 Add PostCreateNamespace hook to E2E tests

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR adds a new PostNamespaceCreated to all E2E tests input objects, thus allowing to inject a function to be run after the test namespace is created. If not specified, this is a no-op.

This can be useful to apply environment settings that apply at the namespace level, eg. setting ResourceQuotas

/area E2E-testing

### CODE REVIEW

File: test/e2e/scale.go [hunk 1]
```diff
@@ -168,6 +172,11 @@ func scaleSpec(ctx context.Context, inputGetter func() scaleSpecInput) {
 			LogFolder:           filepath.Join(input.ArtifactFolder, "clusters", input.BootstrapClusterProxy.GetName()),
 			IgnoreAlreadyExists: true,
 		})
+
+		if input.PostNamespaceCreated != nil {
```
Comment (fabriziopandini) [inline test/e2e/scale.go:176]: Note: this is the only test that is not using setupSpecNamespace (the reason is explained above; we are pinning namespace name vs creating a name with a random suffix).

**Comment by fabriziopandini** [comment]
NOTE: given that the change is clean, non-breaking, and scoped to the test framework only, it would be nice to backport to the 1.6 branch

**Review by chrischdi — COMMENTED** [review]
/lgtm

File: test/e2e/clusterctl_upgrade.go [hunk 2]
```diff
@@ -100,6 +100,9 @@ type ClusterctlUpgradeSpecInput struct {
 	// multiple infrastructure providers (ex: CAPD + in-memory) are installed on the cluster as clusterctl will not be
 	// able to identify the default.
 	InfrastructureProvider *string
+	// Allows to inject a function to be run after test namespace is created.
+	// If not specified, this is a no-op.
+	PostNamespaceCreated func(managementClusterProxy framework.ClusterProxy, workloadClusterNamespace string)
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade.go:105]: Should we call this again in l.351? 

(same in self_hosted.go l.210)

File: test/e2e/clusterctl_upgrade.go [hunk 3]
```diff
@@ -217,7 +217,7 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 		Expect(os.MkdirAll(input.ArtifactFolder, 0750)).To(Succeed(), "Invalid argument. input.ArtifactFolder can't be created for %s spec", specName)
 
 		// Setup a Namespace where to host objects for this spec and create a watcher for the namespace events.
-		managementClusterNamespace, managementClusterCancelWatches = setupSpecNamespace(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder, input.PostNamespaceCreated)
+		managementClusterNamespace, managementClusterCancelWatches = setupSpecNamespace(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder, nil)
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade.go:220]: Don't we want to call it in both places?

File: test/e2e/clusterctl_upgrade.go (hunk 2 above)
Comment (fabriziopandini) [reply test/e2e/clusterctl_upgrade.go:105]: Nice catch, I forgot we have tests who are creating an additional management cluster.
Fixed (we are now calling the hook only for the namespace where we run the test)

For the management cluster itself, I think the correct way to go is what is suggested in https://github.com/kubernetes-sigs/cluster-api/issues/7613, but I would recommend keeping this discussion outside of this PR

File: test/e2e/clusterctl_upgrade.go (hunk 3 above)
Comment (fabriziopandini) [reply test/e2e/clusterctl_upgrade.go:220]: I would say not for now, see https://github.com/kubernetes-sigs/cluster-api/pull/10067#discussion_r1469963592

Comment (sbueringer) [reply test/e2e/clusterctl_upgrade.go:220]: Didn't see your comment. I'll take a look

Comment (fabriziopandini) [reply test/e2e/clusterctl_upgrade.go]: after chatting about this with @sbueringer, in the clusterctl-upgrade and in the self-hosted tests we are calling the func both for the newly created management cluster and the workload cluster

**Comment by fabriziopandini** [comment]
@sbueringer this one should be ready to go now

**Comment by sbueringer** [comment]
Thx!

/lgtm
/approve

**Pull request merged**